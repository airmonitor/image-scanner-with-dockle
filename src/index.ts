import { join } from 'path';
import { CustomResource, Duration } from 'aws-cdk-lib';
import { IRepository } from 'aws-cdk-lib/aws-ecr';
import {
  Architecture,
  AssetCode,
  Handler,
  Runtime,
  SingletonFunction,
} from 'aws-cdk-lib/aws-lambda';
import { Provider } from 'aws-cdk-lib/custom-resources';
import { Construct } from 'constructs';

export interface ImageScannerWithDockleProps {
  readonly imageUri: string;
  readonly repository: IRepository;
  readonly ignore?: string[];
}

export class ImageScannerWithDockle extends Construct {
  constructor(scope: Construct, id: string, props: ImageScannerWithDockleProps) {
    super(scope, id);

    const { imageUri, repository, ignore } = props;

    const customResourceLambda = new SingletonFunction(this, 'CustomResourceLambda', {
      uuid: '662deedd-2d01-1bd0-45de-8d80b1bffe38',
      runtime: Runtime.FROM_IMAGE,
      handler: Handler.FROM_IMAGE,
      code: AssetCode.fromAssetImage(join(__dirname, '../assets/lambda')),
      architecture: Architecture.ARM_64,
      timeout: Duration.seconds(900),
      retryAttempts: 0,
    });
    repository.grantPull(customResourceLambda);

    const imageScannerProvider = new Provider(this, 'Provider', {
      onEventHandler: customResourceLambda,
    });

    const imageScannerProperties: { [key: string]: string | string[] } = {};
    imageScannerProperties.imageUri = imageUri;
    imageScannerProperties.ignore = ignore ?? [];

    new CustomResource(this, 'Default', {
      resourceType: 'Custom::ImageScannerWithDockle',
      properties: imageScannerProperties,
      serviceToken: imageScannerProvider.serviceToken,
    });
  }
}
