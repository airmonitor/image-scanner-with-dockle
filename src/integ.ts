import { resolve } from 'path';
import { App, Stack } from 'aws-cdk-lib';
import { DockerImageAsset, Platform } from 'aws-cdk-lib/aws-ecr-assets';
import { ImageScannerWithDockle } from '.';

const app = new App();
const stack = new Stack(app, 'ImageScannerWithDockleStack');

const image = new DockerImageAsset(stack, 'DockerImage', {
  directory: resolve(__dirname, '../assets/lambda'),
  platform: Platform.LINUX_ARM64,
});

new ImageScannerWithDockle(stack, 'ImageScannerWithDockle1', {
  imageUri: image.imageUri,
  repository: image.repository,
});

new ImageScannerWithDockle(stack, 'ImageScannerWithDockle2', {
  imageUri: image.imageUri,
  repository: image.repository,
  ignore: [],
});

new ImageScannerWithDockle(stack, 'ImageScannerWithDockle3', {
  imageUri: image.imageUri,
  repository: image.repository,
  ignore: ['CIS-DI-0009'],
});
