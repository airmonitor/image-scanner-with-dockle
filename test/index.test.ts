import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { Repository } from 'aws-cdk-lib/aws-ecr';
import { ImageScannerWithDockle } from '../src';

const getTemplate = (): Template => {
  const app = new App();
  const stack = new Stack(app, 'TestStack');

  const repository = new Repository(stack, 'ImageRepository', {});

  new ImageScannerWithDockle(stack, 'ImageScannerWithDockle', {
    imageUri: 'imageUri',
    repository: repository,
    ignore: ['CIS-DI-0009'],
  });
  return Template.fromStack(stack);
};

describe('Fine-grained Assertions Tests', () => {
  const template = getTemplate();

  test('Snapshot test', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });

  test('ImageScannerWithDockle created', () => {
    template.resourceCountIs('Custom::ImageScannerWithDockle', 1);
  });
});
