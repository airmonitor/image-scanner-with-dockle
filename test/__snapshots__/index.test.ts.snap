// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageScannerWithDockle Snapshot test 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CustomImageScannerWithDockleCustomResourceLambda662deedd2d011bd045de8d80b1bffe38E9B33FAD": {
      "DependsOn": [
        "CustomImageScannerWithDockleCustomResourceLambda662deedd2d011bd045de8d80b1bffe38ServiceRoleDefaultPolicyEE41ED61",
        "CustomImageScannerWithDockleCustomResourceLambda662deedd2d011bd045de8d80b1bffe38ServiceRoleA0BDA71E",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "ImageUri": {
            "Fn::Sub": "registry.hub.docker.com/library/busybox",
          },
        },
        "PackageType": "Image",
        "Role": {
          "Fn::GetAtt": [
            "CustomImageScannerWithDockleCustomResourceLambda662deedd2d011bd045de8d80b1bffe38ServiceRoleA0BDA71E",
            "Arn",
          ],
        },
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomImageScannerWithDockleCustomResourceLambda662deedd2d011bd045de8d80b1bffe38EventInvokeConfigFF23ED62": {
      "Properties": {
        "FunctionName": {
          "Ref": "CustomImageScannerWithDockleCustomResourceLambda662deedd2d011bd045de8d80b1bffe38E9B33FAD",
        },
        "MaximumRetryAttempts": 0,
        "Qualifier": "$LATEST",
      },
      "Type": "AWS::Lambda::EventInvokeConfig",
    },
    "CustomImageScannerWithDockleCustomResourceLambda662deedd2d011bd045de8d80b1bffe38ServiceRoleA0BDA71E": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomImageScannerWithDockleCustomResourceLambda662deedd2d011bd045de8d80b1bffe38ServiceRoleDefaultPolicyEE41ED61": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "ecr:BatchCheckLayerAvailability",
                "ecr:GetDownloadUrlForLayer",
                "ecr:BatchGetImage",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "ImageRepositoryBBCBC9DF",
                  "Arn",
                ],
              },
            },
            {
              "Action": "ecr:GetAuthorizationToken",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomImageScannerWithDockleCustomResourceLambda662deedd2d011bd045de8d80b1bffe38ServiceRoleDefaultPolicyEE41ED61",
        "Roles": [
          {
            "Ref": "CustomImageScannerWithDockleCustomResourceLambda662deedd2d011bd045de8d80b1bffe38ServiceRoleA0BDA71E",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ImageRepositoryBBCBC9DF": {
      "DeletionPolicy": "Retain",
      "Type": "AWS::ECR::Repository",
      "UpdateReplacePolicy": "Retain",
    },
    "ImageScannerWithDockle5522CF11": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "ImageScannerWithDockleProviderframeworkonEvent5BEA967D",
            "Arn",
          ],
        },
        "addr": "c8ee6f5f731ae071f1be39c503be0c1f56a08237ff",
        "ignore": [
          "CIS-DI-0009",
        ],
        "imageUri": "imageUri",
      },
      "Type": "Custom::ImageScannerWithDockle",
      "UpdateReplacePolicy": "Delete",
    },
    "ImageScannerWithDockleProviderframeworkonEvent5BEA967D": {
      "DependsOn": [
        "ImageScannerWithDockleProviderframeworkonEventServiceRoleDefaultPolicy6F6DEDE5",
        "ImageScannerWithDockleProviderframeworkonEventServiceRoleF44E92B5",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "HASH-REPLACED.zip",
        },
        "Description": "AWS CDK resource provider framework - onEvent (TestStack/ImageScannerWithDockle/Provider)",
        "Environment": {
          "Variables": {
            "USER_ON_EVENT_FUNCTION_ARN": {
              "Fn::GetAtt": [
                "CustomImageScannerWithDockleCustomResourceLambda662deedd2d011bd045de8d80b1bffe38E9B33FAD",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onEvent",
        "Role": {
          "Fn::GetAtt": [
            "ImageScannerWithDockleProviderframeworkonEventServiceRoleF44E92B5",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ImageScannerWithDockleProviderframeworkonEventServiceRoleDefaultPolicy6F6DEDE5": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "CustomImageScannerWithDockleCustomResourceLambda662deedd2d011bd045de8d80b1bffe38E9B33FAD",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "CustomImageScannerWithDockleCustomResourceLambda662deedd2d011bd045de8d80b1bffe38E9B33FAD",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ImageScannerWithDockleProviderframeworkonEventServiceRoleDefaultPolicy6F6DEDE5",
        "Roles": [
          {
            "Ref": "ImageScannerWithDockleProviderframeworkonEventServiceRoleF44E92B5",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ImageScannerWithDockleProviderframeworkonEventServiceRoleF44E92B5": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
